jgrapht-core/src/main/java/org/jgrapht/alg/tour/GreedyHeuristicTSP.java

```diff
@@ -0,0 +1,139 @@
+/*
+ * (C) Copyright 2019-2019, by Peter Harman and Contributors.
+ *
+ * JGraphT : a free Java graph-theory library
+ *
+ * See the CONTRIBUTORS.md file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the
+ * GNU Lesser General Public License v2.1 or later
+ * which is available at
+ * http://www.gnu.org/licenses/old-licenses/lgpl-2.1-standalone.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR LGPL-2.1-or-later
+ */
+package org.jgrapht.alg.tour;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Deque;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.jgrapht.Graph;
+import org.jgrapht.GraphPath;
+import org.jgrapht.graph.MaskSubgraph;
+import org.jgrapht.traverse.DepthFirstIterator;
+
+/**
+ * The greedy heuristic algorithm for the TSP problem.
+ *
+ * <p>
+ * The travelling salesman problem (TSP) asks the following question: "Given a
+ * list of cities and the distances between each pair of cities, what is the
+ * shortest possible route that visits each city exactly once and returns to the
+ * origin city?".
+ * </p>
+ *
+ * <p>
+ * The Greedy heuristic gradually constructs a tour by repeatedly selecting the
+ * shortest edge and adding it to the tour as long as it doesn’t create a cycle
+ * with less than N edges, or increases the degree of any node to more than 2.
+ * We must not add the same edge twice of course.
+ * </p>
+ *
+ * <p>
+ * The implementation of this class is based on: <br>
+ * <a href="http://160592857366.free.fr/joe/ebooks/ShareData/Heuristics%20for%20the%20Traveling%20Salesman%20Problem%20By%20Christian%20Nillson.pdf">Nilsson
```

Comment:
Please add a proper reference:
Nilsson, Christian. "Heuristics for the traveling salesman problem." Linkoping University 38 (2003)